### ISSUE TITLE
oc binary not in path

### ISSUE THREAD

**Issue by kwoodson** [openshift/openshift-ansible#3410]
Provide a brief description of your issue here. For example:

On certain installs such as AH, or by running as different users in the ansible inventory the oc binary is not in the PATH.  This is causing errors when attempting to run the modules:

`FileNotFoundError: [Errno 2] No such file or directory: 'oc'\n"`

##### Version

ansible 2.2

##### Steps To Reproduce
1. In the inventory file set your user to vagrant, ec2_user, or a user whose path does not have oc
2. Run any of the ansible lib_openshift modules.

##### Expected Results

I would expect that the modules run successfully.

##### Observed Results
When oc does not exist on the path the user sees: 
```
FileNotFoundError: [Errno 2] No such file or directory: 'oc'\n"
```

##### Additional Information

Currently the expectation for the oc binary is to live in the path of the user.  When that path does not define /usr/local/bin or any other path to where the binary is currently installed then we receive failures.

I think we can make an attempt at being a bit more resilient.  We can try and find the oc binary in the path, and if that fails, we can attempt a few "known" locations.
- ~/bin/
- /usr/local/bin

**Comment by dustymabe** [comment]
yeah the key to this is that `/usr/local/bin/` isn't in the $PATH when running `sudo` so it doesn't work with some random user and `ansible_become=true`.

**Comment by sdodson** [comment]
@thoraxe is the previous comment what you're running into?

**Comment by detiber** [comment]
We can probably leverage the `get_bin_path` method from AnsibleModule to help:
```
oc_bin = module.get_bin_path('oc', opt_dirs=['/usr/local/bin', os.path.expanduser('~/bin/oc')])
```

**Comment by cgwalters** [comment]
I'll work on this.

**Comment by cgwalters** [comment]
Unless you are @kwoodson ?

**Comment by cgwalters** [comment]
PR in https://github.com/openshift/openshift-ansible/pull/3413

**Comment by thoraxe** [comment]
@sdodson if going sudo to root counts (as opposed to running a command with sudo) then yes I suppose it's the same problem. I didn't actually try to su to root.

**Comment by rhcarvalho** [comment]
> ~/bin/

@kwoodson in this case, `~` should expand relative to what user?


---

Perhaps instead of changing the PATH of arbitrary users, or finding executables outside of PATH, wouldn't it be possible to arrange the playbooks in such a way that `oc`, etc, is in the PATH? I think that would be less puzzling, as it corresponds to how Linux works.

Imagine a potential future bug report when the version of `oc` being used is wrong because we're changing the PATH under the covers...

**Comment by detiber** [comment]
> Perhaps instead of changing the PATH of arbitrary users, or finding executables outside of PATH, wouldn't it be possible to arrange the playbooks in such a way that oc, etc, is in the PATH?

The problem is that for containerized installs we copy the client from the container to /usr/local/bin to avoid conflicting with the rpm managed binary, which we should prefer when deploying to a mixed environment (think master containerized, node not on same host).

The main issue, though, is that sudo does not pass through the PATH environment variable in the default config. This leaves us with a bunch of suboptimal choices:
- configure sudo to pass the PATH env var, which is a non-starter imo.
- configure the user's PATH in a persistent way, which I would also consider a non-starter.
- set the PATH env var on tasks that need it, which is going to be repetitive and error prone.
- use the full binary path to the client, which would need to set it properly according to containerized/non-containerized, etc. I worry this is going down the path of black-box/openshift_facts, which we are trying to get away from because of the complexity and difficulty for users to understand.
- check for known paths after doing a normal path lookup, not ideal, but it is the least surprising/magical I can think of.

One other thing to note is that the modules can be used outside of the playbooks we ship and could be used in dev/test envs that do not have the RPMs installed and wouldn't have permissions to copy the client binary to /usr/local/bin. Think `oc cluster up`. ~/bin is a common location for those cases, and not all users may think/know to alter their PATH.

>  I think that would be less puzzling, as it corresponds to how Linux works.

I mostly agree, which is why we should rely on the default behavior as much as possible, but I think falling back to known locations is prudent as well, since it will handle the issue at hand, improve dev/test experience and also work user-environment issues we've seen in the past with customized environments (policy-enforced env var settings, etc).

> Imagine a potential future bug report when the version of oc being used is wrong because we're changing the PATH under the covers.

Agreed. We'll have to support some way of handling this, the question is which is the least worst approach.

**Comment by rhcarvalho** [comment]
@detiber would it be correct than to summarize as:

- First, try to find `oc` (or `oadm` or anything else) in $PATH.
- If found, use it.
- If not found, then search it on `/usr/local/bin` and then `~/bin`

?

That sounds easy to explain in docs to me. I'm still not clear on what user's HOME should `~` expand to.

**Comment by detiber** [comment]
> @detiber would it be correct than to summarize as:
> - First, try to find oc (or oadm or anything else) in $PATH.
> - If found, use it.
> - If not found, then search it on /usr/local/bin and then ~/bin

That is the behavior I would expect. I would add one other item as a fallback:
- If not found after searching /usr/local/bin or ~/bin, then just attempt to execute 'oc' without the path specified.

That would allow us to fallback to allowing the system to try and find the binary if we were not able to find it.

For ~/bin I would expect to to expand to the user executing the module, so depending on the become setting:
- become: no -> $HOME for ssh_user
- become: yes -> $HOME for sudo user

**Comment by thoraxe** [comment]
Is there a particular reason we are putting this in `/usr/local/bin` instead of `/usr/local/sbin` which is in `root`s path when using sudo?

**Comment by cgwalters** [comment]
@thoraxe Not here:

```
[vagrant@localhost ~]$ atomic host status
State: idle
Deployments:
● centos-atomic-continuous:centos-atomic-host/7/x86_64/devel/alpha
             Version: 7.2017.109 (2017-02-16 15:36:48)
          BaseCommit: 71f0d9f2784d19ac74e4c3bd6cffac20ae31f23c4b8e847b40011ec4aa95dcb5
              Commit: ddf91052764499c50caa66666fff7c0495b416f94c922648df991b4a13765dc5
              OSName: centos-atomic-host
            Packages: python2-ruamel-yaml
[vagrant@localhost ~]$ sudo env|grep PATH
PATH=/sbin:/bin:/usr/sbin:/usr/bin
[vagrant@localhost ~]$ 
```

**Comment by thoraxe** [comment]
So then perhaps we should be putting it in `/usr/sbin` ?

It seems like we are trying to go the long way around in order to get the file into the path...

**Comment by detiber** [comment]
@thoraxe I'm not sure we'd want to limit oc to "admin" users anyway.

**Comment by dustymabe** [comment]
`/usr` is a read only filesystem on some platforms (i.e. Atomic Host) by design.

**Comment by rhcarvalho** [comment]
> I would add one other item as a fallback:
> 
> - If not found after searching /usr/local/bin or ~/bin, then just attempt to execute 'oc' without the path specified.

Implementation-wise, that's exactly what `subprocess.Popen` does when we specify a name and not a full path.
Thus my suggestion in the PR to augment the PATH and simply call `oc` as it is in master.

**Comment by lmello** [comment]
This needs to be merged to release-1.5 branch also.
Tried to install Openshift version 1.5 alpha.3 with release-1.5 branch and it failed because of this issue.
